fix(docs): Update the token bridge tutorial (#11578)
AztecProtocol/aztec-packages: docs/docs/developers/tutorials/codealong/contract_tutorials/token_bridge.md

  - `Token`: Manages the tokens on L2
  - `TokenBridge`: Manages the bridging of tokens between L2 and L1

`TokenPortal.sol` is the contract that manages the passing of messages from L1 to L2. It is deployed on L1, is linked to a specific token on L1 and a corresponding contract on L2. The `registry` is used to find the rollup and the corresponding `inbox` and `outbox` contracts.

## How it works

### Deposit to Aztec

`TokenPortal.sol` passes messages to Aztec both publicly and privately.

Message content that is passed to Aztec is limited to a single field element (~254 bits), so if the message content is larger than that, it is hashed, and the message hash is passed an verified on the receiving contract. There is a utility function in the `Hash` library to hash messages (using `sha256`) to field elements.

The Aztec message Inbox expects a recipient Aztec address that can consume the message (the corresponding L2 bridge contract), the Aztec version (similar to Ethereum's `chainId`), the message content hash (which includes the token recipient and amount in this case), and a `secretHash`, where the corresponding `secret` is used to consume the message on the receiving contract.

> So in summary, it deposits tokens to the portal, encodes a mint message, hashes it, and sends it to the Aztec rollup via the Inbox. The L2 token contract can then mint the tokens when it processes the message.

Author: wdyt? 
![image](https://github.com/user-attachments/assets/1ec7a19e-cbe1-4765-a4be-5279cfaa38ac)

Note that because L1 is public, everyone can inspect and figure out the contentHash and the recipient contract address.

> #### `depositToAztecPublic`

Reviewer: ```suggestion
#### `depositToAztecPublic` (TokenPortal.sol)
```

#include_code deposit_public l1-contracts/test/portals/TokenPortal.sol solidity

> #### `depositToAztecPrivate`

Reviewer: ```suggestion
#### `depositToAztecPrivate` (TokenPortal.sol)
```

### Minting on Aztec

In the previous step, we moved our funds to the bridge and created a L1->L2 message. Upon building the next rollup block, the sequencer asks the L1 inbox contract for any incoming messages and adds them to the Aztec block's L1->L2 message tree, so an application on L2 can prove that the message exists and can consume it.

#include_code claim_public /noir-projects/noir-contracts/contracts/token_bridge_contract/src/main.nr rust

What's happening here?

1. compute the content hash of the message
2. consume the message
3. mint the tokens

:::note

The Aztec `TokenBridge` contract should be an authorized minter in the corresponding Aztec `Token` contract so that it is able to complete mints to the intended recipient.

:::

### Withdraw to L1

> Now we have tokens on L2, we can withdraw them back to L1. You can withdraw part of a public or private balance to L1, but the amount and the recipient on L1 will be public.

Reviewer: ```suggestion
The token bridge also allows tokens to be withdrawn back to L1 from L2. You can withdraw part of a public or private balance to L1, but the amount and the recipient on L1 will be public.
```

#### `exit_to_L1_private`

This function works very similarly to the public version, except here we burn user’s private notes.

#include_code exit_to_l1_private /noir-projects/noir-contracts/contracts/token_bridge_contract/src/main.nr rust

Since this is a private method, it can't read what token is publicly stored. So instead the user passes a token address, and `_assert_token_is_same()` checks that this user provided address is same as the one in storage.

Because public functions are executed by the sequencer while private methods are executed locally, all public calls are always done _after_ all private calls are completed. So first the burn would happen and only later the sequencer asserts that the token is same. The sequencer just sees a request to `execute_assert_token_is_same` and therefore has no context on what the appropriate private method was. If the assertion fails, then the kernel circuit will fail to create a proof and hence the transaction will be dropped.

A user must sign an approval message to let the contract burn tokens on their behalf. The nonce refers to this approval message.

> ### L1 `TokenPortal.sol`

Reviewer: ```suggestion
### Claiming on L1
```

After the transaction is completed on L2, the portal must call the outbox to successfully transfer funds to the user on L1. Like with deposits, things can be complex here. For example, what happens if the transaction was done on L2 to burn tokens but can’t be withdrawn to L1? Then the funds are lost forever! How do we prevent this?

> #include_code token_portal_withdraw /l1-contracts/test/portals/TokenPortal.sol solidity

Reviewer: ```suggestion
#### token_portal_withdraw (TokenPortal.sol)
#include_code token_portal_withdraw /l1-contracts/test/portals/TokenPortal.sol solidity
```

> Here we reconstruct the L2 to L1 message and check that this message exists on the outbox. If so, we consume it and transfer the funds to the recipient. As part of the reconstruction, the content hash looks similar to what we did in our bridge contract on aztec where we pass the amount and recipient to the hash. This way a malicious actor can’t change the recipient parameter to the address and withdraw funds to themselves.

Reviewer: ```suggestion
Here we reconstruct the L2 to L1 message and check that this message exists on the outbox. If so, we consume it and transfer the funds to the recipient. As part of the reconstruction, the content hash looks similar to what we did in our bridge contract on Aztec where we pass the amount and recipient to the hash. This way a malicious actor can’t change the recipient parameter to the address and withdraw funds to themselves.
```

We also use a `_withCaller` parameter to determine the appropriate party that can execute this function on behalf of the recipient. If `withCaller` is false, then anyone can call the method and hence we use address(0), otherwise only msg.sender should be able to execute. This address should match the `callerOnL1` address we passed in aztec when withdrawing from L2.

We call this pattern _designed caller_ which enables a new paradigm **where we can construct other such portals that talk to the token portal and therefore create more seamless crosschain legos** between L1 and L2.

## Running with Aztec.js

Let's run through the entire process of depositing, minting and withdrawing tokens in Typescript, so you can see how it works in practice.

### Prerequisites

Same prerequisites as the [getting started guide](../../../../developers/getting_started.md#prerequisites) and the sandbox.

### ProjectSetup

Create a new directory for the tutorial and install the dependencies:

```bash
mkdir token-bridge-tutorial
cd token-bridge-tutorial
yarn init -y

> yarn add @aztec/aztec.js @aztec/noir-contracts.js @aztec/l1-artifacts @aztec/accounts @aztec/ethereum @aztec/types @types/node typescript@^5.0.4 viem@^2.22.8 tsx

Reviewer: in new aztec sandbox i also had to add tsx and lodash.omit
```suggestion
yarn add @aztec/aztec.js @aztec/noir-contracts.js @aztec/l1-artifacts @aztec/accounts @aztec/ethereum @aztec/types @types/node typescript@^5.0.4 viem@^2.22.8 tsx tslib lodash.omit
```

Author: Can you share the error you see without these installed? I just ran through this again and it works for me without these additional packages. Maybe its a mac vs linux thing?

Author: @signorecello was hitting some errors as well, installing node versino 18.19 made it work. I will update the file to specify the node version

Reviewer: this is the version i am using too. the lodash error was `Error: @aztec/circuit-types tried to access lodash.omit, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.` and tslib was the usual cannot find

Author: I found the problem. Can you create .yarnrc.yml and specify the node linker? 

```
echo "nodeLinker: node-modules" > .yarnrc.yml
```

Reviewer: this works

### Claim on Aztec

Add the following code to `index.ts` to claim the tokens publicly on Aztec:

#include_code claim /yarn-project/end-to-end/src/e2e_token_bridge_tutorial_test.test.ts typescript

### Withdraw

Add the following code to `index.ts` to start the withdraw the tokens to L1:

#include_code setup-withdrawal /yarn-project/end-to-end/src/e2e_token_bridge_tutorial_test.test.ts typescript

We have to send a public authwit to allow the bridge contract to burn tokens on behalf of the user.

Add the following code to `index.ts` to start the withdraw process on Aztec:

#include_code l2-withdraw /yarn-project/end-to-end/src/e2e_token_bridge_tutorial_test.test.ts typescript

Add the following code to `index.ts` to complete the withdraw process on L1:

#include_code l1-withdraw /yarn-project/end-to-end/src/e2e_token_bridge_tutorial_test.test.ts typescript

> Run `yarn start` to run the script and see the entire process in action.

Reviewer 2: General point: the dev is doing mostly copy/pasting before being invited to run something smaller and test the output. I suppose they could do it themselves, but there might be a few points in the flow where a one-line would help understanding key things along the way. Eg: "Try this with `yarn start`, and print the l1 address"

Author: I added a few reminders at the bottom of sections to run this
